### ISSUE TITLE
Deprecating uses of a type or constant

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#11043]
The `@[Deprecated]` annotation has no effect when applied to a type:

```crystal
@[Deprecated]
class Foo
end

Foo.new # no warnings

class Bar < Foo # no warnings
end

alias Baz = Foo | Int32 # no warnings

def foo(x : Foo) # no warnings
end
```

All of these are uses of `Foo` and should generate warnings. Any constant that resolves to `Foo` is considered a use of `Foo` (because constants have the same lookup rules as type names, it follows that they should support the same deprecation mechanism). This is distinct from #10125 which is only concerned with aliases.

**Comment by straight-shoota** [comment]
I don't think `def foo(x : Foo)` should count as a use on its own. A deprecated method only creates a warning, if it is called. It should work similarly for deprecated types in a method's type restrictions.

**Comment by ysbaddaden** [comment]
A couple more cases:

```crystal
@[Deprecated]
module XXX
end

class ZZZ
  include XXX # no warning
  extend XXX # no warning
end
```

**Comment by straight-shoota** [comment]
More use cases:
Paths into a deprecated namespace.

```cr
@[Deprecated]
class Foo
  BAZ = 1

  class Qux
  end
end

Foo::BAZ # no warning
Foo::Qux # no warning
```

There might be some use case for keeping `Foo` as a namespace only where you would only want to deprecate using it directly as a type. However, a namespace could also contain class methods, which trigger a deprecation warning with #15962.
So this wouldn't really work well. I believe the correct path for this would be deprecating the individual `def`s for constructors and instance methods.

Optionally, we could consider an option to the `@[Deprecated]` annotation to specify whether the deprecation affects everything within that namespace or not. But this is probably not necessary.

**Comment by ysbaddaden** [comment]
> I believe the correct path for this would be deprecating the individual defs for constructors and instance methods.

I agree with that. The "type" itself isn't deprecated if you can still legitimately reference it.

I suppose a deprecated type should affect everything inside the namespace (similar to `:nodoc:`): class methods, methods and nested types and constants... That makes things a bit more complex, because we must resolve each type in the `Path` (`Foo::Bar` then `Foo`), not just the last (`Foo::Bar`). Maybe we should copy the deprecated annotation onto nested types?

The alternative is to not affect the namespace, and expect an explicit deprecated annotation on nested types.

**Comment by ysbaddaden** [comment]
I was wondering if instance methods should also be deprecated, but maybe not: the constructors are already deprecated so it might just be noise; same for modules, as long as we warn on `include` and `extend` then it'd just be noise.

Docs might consider this differently.

**Comment by straight-shoota** [comment]
Another use case is references in a generic argument:

```cr
@[Deprecated]
class Foo
end

class Bar(T)
end

Bar(Foo) # should warn
```
